server/metrics/global_status.go

```diff
@@ -0,0 +1,216 @@
+		timeChunks: timeChunksMap,
+		activeChunk: &TenantStatusTimeChunk{
+			startTime: startTime,
+			tenants:   tenantMap,
+		},
+	}
+}
+
+func NewTenantStatusTimeChunk(startTime time.Time) *TenantStatusTimeChunk {
+	tenantMap := make(map[string]*TenantStatus)
+	return &TenantStatusTimeChunk{tenants: tenantMap, startTime: startTime}
+}
+
+func NewTenantStatus() *TenantStatus {
+	return &TenantStatus{}
+}
+
+func RequestStatusFromContext(ctx context.Context) (*RequestStatus, bool) {
+	r, ok := ctx.Value(RequestStatusCtxKey{}).(*RequestStatus)
+	return r, ok
+}
+
+func (r *RequestStatus) SaveRequestStatusToContext(ctx context.Context) context.Context {
+	ctx = context.WithValue(ctx, RequestStatusCtxKey{}, r)
+	return ctx
+}
+
+func (r *RequestStatus) ClearRequestStatusFromContext(ctx context.Context) context.Context {
+	return context.WithValue(ctx, RequestStatusCtxKey{}, nil)
+}
+
+func (r *RequestStatus) CountReadsOff() {
+	r.countReads = false
+}
+
+func (r *RequestStatus) AddReadBytes(value int64) {
+	if r.countReads {
+		r.readBytes += value
+		log.Debug().Int64("Bytes read", value).Int64("Total bytes read", r.readBytes).Msg("Added read bytes")
+	}
+}
+
+func (r *RequestStatus) AddWriteBytes(value int64) {
+	if r.countWrites {
+		r.writeBytes += value
+		log.Debug().Int64("Bytes written", value).Int64("Total bytes written", r.writeBytes).Msg("Added written bytes")
+	}
+}
+
+func (r *RequestStatus) SetWriteBytes(value int64) {
+	r.writeBytes = value
+}
+
+func (r *RequestStatus) SetReadBytes(value int64) {
+	r.readBytes = value
+}
+
+func (r *RequestStatus) GetReadBytes() int64 {
+	return r.readBytes
+}
+
+func (r *RequestStatus) GetWriteBytes() int64 {
+	return r.writeBytes
+}
+
+func (g *GlobalStatus) MergeActiveChunk() {
+	endTime := time.Now()
+	g.mu.Lock()
+	activeChunk := *g.activeChunk
+	log.Debug().Time("start time", activeChunk.startTime).Time("end time", endTime).Msg("Merging active chunk")
+	activeChunk.endTime = endTime
+	g.timeChunks[activeChunk.startTime] = &activeChunk
+	newChunk := NewTenantStatusTimeChunk(endTime)
+	g.activeChunk = newChunk
+	g.mu.Unlock()
+	g.LogChunks()
+}
+
+func (g *GlobalStatus) LogChunks() {
+	for tsidx, chunk := range g.timeChunks {
+		log.Debug().Time("tsidx", tsidx).Time("start time", chunk.startTime).Time("end time", chunk.endTime).Int("tenants length", len(chunk.tenants)).Msg("chunk metadata")
+		for tenantName, tenantStatus := range chunk.tenants {
+			log.Debug().Str("tenant name", tenantName).Int64("readBytes", tenantStatus.readBytes).Int64("writeBytes", tenantStatus.writeBytes).Msg("Bytes for tenant")
+			log.Debug().Str("tenant name", tenantName).Int64("writeUnits", tenantStatus.writeUnits).Int64("readUnits", tenantStatus.readUnits).Msg("Units for tenant")
+		}
+	}
+}
+
+func (g *GlobalStatus) GetActiveChunk() *TenantStatusTimeChunk {
+	return g.activeChunk
+}
+
+func (g *GlobalStatus) EnsureTenantForActiveChunk(tenantName string) {
+	_, ok := g.activeChunk.tenants[tenantName]
```

Comment:
If these methods are exposed then don't we need to lock the tenants map to avoid any concurrent write happening to it?

Reply:
With moving to flush, this became obsolete.